WLR Layer Shell Support

PolyMeilex opened the pull request:
Design goal of this implementation was to make its API as similar to `xdg-shell` as possible, the idea is that if someone already knows how to use `xdg-shell` they should be able to start using `layer-shell` by using nearly the same API. So `CachedState`, `with_pending_state`, `current_state`, etc. are all there, and should act just like xdg equivalents.


Anvil implementation of layer shell is lacking at the moment, I wanted to implement everything but the more I added, the hackier the code felt, so I decided to stop. The main missing piece for anvil is exclusive_zone support, but I think that I implemented enough to make it a useful test ground for future changes.
 

https://user-images.githubusercontent.com/20758186/127673456-3296c042-69a6-478f-8a71-dccf66dbef8b.mp4

vberger commented:
So, a few minor points, mostly regarding documentation. Other than that this looks pretty solid, thank! :+1:

PolyMeilex commented:
All long lines got eradicated.
Changelog is now fixed and flows the convention.
I rebased on top of master.

While at it, I also exposed the surface role `&str` just like #342 does for other shells

i509VCB commented on src/wayland/shell/wlr_layer/mod.rs:
```diff
@@ -0,0 +1,454 @@
+pub use types::{Anchor, ExclusiveZone, KeyboardInteractivity, Layer, Margins};
+
+/// The role of a wlr_layer_shell_surface
+pub static LAYER_SURFACE_ROLE: &str = "zwlr_layer_surface_v1";
```
Probably a little better to make this `pub const` here. Static associates a memory address with the value and using `const` allows the values to be inlined a little more easily by rustc.

PolyMeilex replied:
Good point, it's better to use compile time constants for this
